src/include/rados_unleashed/rados_unleashed.hpp

```diff
@@ -0,0 +1,604 @@
+
+  void get_omap_keys(std::optional<std::string_view> start_after,
+		     uint64_t max_return);
+  // → std::pair<boost::container::flat_set<std::string>, bool>
+
+
+  void get_xattrs();
+  // → boost::container::flat_map<std::string, ceph::buffer::list>;
+
+  struct GetOmapValsRes {
+  };
+  void get_omap_vals(std::optional<std::string_view> start_after,
+		     std::optional<std::string_view> filter_prefix,
+		     uint64_t max_return);
+  // → std::pair<boost::container::flat_map<std::string, ceph::buffer::list>,
+  //                                        bool>
+
+  void get_omap_vals_by_keys(const boost::container::flat_set<std::string>& keys);
+  // → boost::container::flat_map<std::string, ceph::buffer::list>;
+
+  void list_watchers(); // → std::vector<obj_watch_t>
+  void list_snaps(); // → librados::snap_set_t
+
+  void exec(std::string_view cls, std::string_view method,
+	    const bufferlist& inbl); // → ceph::buffer::list
+
+private:
+  ReadOp();
+};
+
+class WriteOp : public Op {
+  friend RADOS;
+public:
+
+  using Op::Op;
+  using Op::operator =;
+
+  WriteOp(const WriteOp&) = delete;
+  WriteOp& operator =(const WriteOp&) = delete;
+  WriteOp(WriteOp&&);
+  WriteOp& operator =(WriteOp&&);
+
+  ~WriteOp();
+
+  void set_mtime(ceph::real_time t);
+  void create(bool exclusive);
+  void write(uint64_t off, bufferlist&& bl);
+  void write_full(bufferlist&& bl);
+  void writesame(std::uint64_t off, std::uint64_t write_len,
+		 bufferlist&& bl);
+  void append(bufferlist&& bl);
+  void remove();
+  void truncate(uint64_t off);
+  void zero(uint64_t off, uint64_t len);
+  void rmxattr(std::string_view name);
+  void setxattr(std::string_view name,
+		bufferlist&& bl);
+  void rollback(uint64_t snapid);
+  void set_omap(const boost::container::flat_map<std::string, bufferlist>& map);
+  void set_omap_header(bufferlist&& bl);
+  void clear_omap();
+  void rm_omap_keys(const boost::container::flat_set<std::string>& to_rm);
+  void set_alloc_hint(uint64_t expected_object_size,
+		      uint64_t expected_write_size,
+		      uint32_t flags);
+  void exec(std::string_view cls, std::string_view method,
+	    const bufferlist& inbl);
+private:
+  WriteOp();
+};
+
+class RADOS
+{
+public:
+  static constexpr std::tuple<uint32_t, uint32_t, uint32_t> version() {
+    return {0, 0, 1};
+  }
+
+  RADOS(boost::asio::io_context& ioctx);
+  RADOS(boost::asio::io_context& ioctx, std::string_view id);
+  RADOS(boost::asio::io_context& ioctx, std::string_view name,
+	std::string_view cluster);
+  RADOS(boost::asio::io_context& ioctx, CephContext* cct);
+
+  RADOS(const RADOS&) = delete;
+  RADOS& operator =(const RADOS&) = delete;
+  RADOS(RADOS&&) = delete;
+  RADOS& operator =(RADOS&&) = delete;
+
+  CephContext* cct();
+
+  ReadOp make_ReadOp();
+  WriteOp make_WriteOp();
+
+  template<typename Executor, typename CompletionToken>
+  auto execute(const Object& o, const IOContext& ioc, ReadOp&& op,
+	       Executor&& ex, CompletionToken&& token) {
```

Comment:
Executor doesn't need to be part of the interface, because the CompletionToken already allows the caller to customize this - to run some `handler` on an executor `ex`, you can pass `boost::asio::bind_executor(ex, std::move(handler))` as the token

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/n4771.pdf#section.13.2
Section 13.2.7 "Requirements on asynchronous operations" is a really good read on how this stuff works

> 13.2.7.8.2.1 — If the initiating function is a member function, the associated executor is that returned by the `get_executor` member function on the same object.

so RADOS should also provide a `get_executor()` member function:
```c++
using executor_type = boost::asio::io_context::executor_type;
executor_type get_executor() noexcept { return ioctx.get_executor(); } // except through _::RADOS
```

then we can pass `get_executor()` as the first argument to `Completion::create()` so it's used whenever the provided Handler doesn't already have an [associated executor](https://www.boost.org/doc/libs/1_68_0/doc/html/boost_asio/reference/get_associated_executor.html) - whether that's from bind_executor() or some other specialization of `boost::asio::associated_executor<Handler>`